CalendarApp/Logic Handlers/EventSyncHandler.m

```diff
@@ -0,0 +1,146 @@
+        self.context = ((AppDelegate *)UIApplication.sharedApplication.delegate).persistentContainer.viewContext;
+        self.parseEventHandler = [[ParseEventHandler alloc] init];
+        self.coreDataEventHandler = [[CoreDataEventHandler alloc] init];
+        self.userData = NSUserDefaults.standardUserDefaults;
+        
+        self.networkHandler = [[NetworkHandler alloc] init];
+        self.networkHandler.delegate = self;
+        [self.networkHandler startMonitoring];
+    }
+    return self;
+}
+
+- (void)didChangeOnline {
+    if (self.isSynced) {
+        return;
+    }
+    self.isSynced = true;
+    if (![self.userData objectForKey:@"lastUpdated"]) {
+        // TODO: prompt user to query all and sync with all existing remote events
+        [self.userData setObject:[NSDate date] forKey:@"lastUpdated"];
+        [self.userData synchronize];
+        return;
+    }
+    [self.parseEventHandler queryEventsAfterUpdateDate:[self.userData objectForKey:@"lastUpdated"]
+                                            completion:^(BOOL success,
+                                                         NSMutableArray<Event *> * _Nullable events,
+                                                         NSDate * _Nullable updatedDate,
+                                                         NSString * _Nullable error) {
+        if (!success) {
+            // TODO: Error handling
+        } else {
+            NSArray<LocalChange *> *localChanges = [self.context executeFetchRequest:LocalChange.fetchRequest error:nil];
+            
+            SyncConflictHandler *conflictHandler = [[SyncConflictHandler alloc] init];
+            NSArray<LocalChange *> *keptChanges = [conflictHandler getChangesToSyncWithOnlineEvents:events offlineChanges:localChanges];
+            [self syncLocalChanges:keptChanges];
+            [self deleteAllLocalChanges];
+            [self.userData setObject:[NSDate date] forKey:@"lastUpdated"];
+        }
+    }];
+}
+
+- (void)deleteAllLocalChanges {
+    NSBatchDeleteRequest *delete = [[NSBatchDeleteRequest alloc] initWithFetchRequest:LocalChange.fetchRequest];
+    NSPersistentStoreCoordinator *persistentStoreCoordinator = ((AppDelegate *)UIApplication.sharedApplication.delegate).persistentContainer.persistentStoreCoordinator;
+    [persistentStoreCoordinator executeRequest:delete withContext:self.context error:nil];
+}
+
+- (void)syncLocalChanges:(NSArray<LocalChange *> *)localChanges {
+    for (LocalChange *change in localChanges) {
+        [self syncEventToParse:change.oldEvent updatedEvent:change.updatedEvent];
+    }
+}
+
+- (void)didChangeOffline {
+    self.isSynced = false;
+}
+
+- (void)syncEventToParse:(Event *)oldEvent
+            updatedEvent:(Event *)newEvent {
+    if (!oldEvent) {
+        [self syncNewEventToParse:newEvent];
+    } else if (!newEvent) {
+        [self syncDeleteToParse:oldEvent];
+    } else {
+        [self syncUpdateToParse:newEvent];
+    }
+}
```

Comment:
This is another good example of how testing could potentially drive the design. If I wanted to test this I would probably first write a type like,
```
enum Change {
    case create(Event)
    case delete(Event)
    case update(Event)
}
```
Then you could just have a method that takes two events and returns the change to use. This is more testable since you have outputs instead of side effects.